Semi-formal specification for method resolution

Akirathan opened the pull request:
### Pull Request Description

Semi-formal spec for method invocation. Documents the **desired behavior** that is being implemented in #13962.

### Important Notes

<!--
- Mention important elements of the design.
- Mention any notable changes to APIs.
-->

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [x] Unit tests have been written where possible.
- [x] If meaningful changes were made to logic or tests affecting Enso Cloud integration in the libraries, 
      or the Snowflake database integration, a run of the [Extra Tests](https://github.com/enso-org/enso/actions/workflows/extra-nightly-tests.yml) has been scheduled.
  - If applicable, it is suggested to paste a link to a successful run of the Extra Tests.

Commit:
docs for method resolution

Akirathan commented:
RFC engine and libs team (@JaroslavTulach , @jdunkerley , @4e6 , @hubertp , @GregoryTravis ): Do you consider this kind of semi-formal specification for method resolution and method evaluation useful? Is it better than looking at our tests or in our code base? Do you want me to finish it? Do you have any other suggestions how to improve it? Note that this PR is WIP, obviously.

JaroslavTulach commented on docs/types/dynamic-dispatch.md:
```diff
@@ -162,3 +162,137 @@ main = Test.simplify
+
+5. **`Receiver` is a polyglot object. A polyglot object can be:**
+
+   - 5.1. A Java class, imported by `polyglot java import ...` statement.
```
- handling polyglot is very different special case.
- as Ara puts it: it is a [generic variadic syntax](https://github.com/enso-org/enso/blob/27310706afc8dc7a5a52d1c758e08efcab62d3d8/docs/polyglot/polyglot-bindings.md?plain=1#L60) 
   - e.g. collect all arguments and send them all into `invokeMember`
- there are no pools to "lookup" methods at
- e.g. just reference the other document than trying to describe the dispatch here

Akirathan replied:
Done in 5f6fdaa71cccb998f17f311e68d711c21d45d46a

JaroslavTulach commented on docs/types/dynamic-dispatch.md:
```diff
@@ -162,3 +162,137 @@ main = Test.simplify
+   - 1.4. If `Receiver` is a polyglot object, go to step 5.
+   - 1.5. If there is no `Receiver`, go to step 6.
+
+2. **`Receiver` is a type. This call site is a _static method call_:**
```
To play devil's advocate: 
- Isn't it all a bit different?
- The 1st step is to find a `Type` for the receiver
   - for `Atom` the type is its constructor defining type
   - for module the type is its associated type
   - for a `Type` it is its _eigen type_
   - etc.?
- then lookup the method in the _pool_ of that type or
   - its `getSupertype` pool and so on
- if a method is found, invoke it with `self` being set to the receiver
- 00577f7c96 documents/tests the _chain of types_ and shows how we want to use `allTypes` method
- if no such a method is found, consider _static invocation of instance method_!?


_and somewhere here my thoughts dry out..._

JaroslavTulach replied:
> _and somewhere here my thoughts dry out..._
> * [00577f7](https://github.com/enso-org/enso/commit/00577f7c9664f6f21bcc1727dcf09e4c6f75507a) documents/tests the _chain of types_ and shows how we want to use `allTypes` method

- I am still relatively satisfied with 00577f7
   - e.g. there are no "static" methods - those are instance methods on `Type.getEigentype()`
- the only problem is the "duality of methods from Any"
- and I believe it is sufficiently solved by saying that:
   - each method found on `Any` has `self` argument defaulted to the receiver
   - but it can be changed by `self=whatever`
- at the end this sounds like a consistent set of rules

Akirathan replied:
I have simplified the Method resolution specs in 5f6fdaa71cccb998f17f311e68d711c21d45d46a according to your suggestions, which makes sense:
1. Determine type of the `Receiver`
2. Symbol lookup on the determined type.

JaroslavTulach committed:
Testing/documenting the type chains for Atom, Type and Module

JaroslavTulach commented on docs/types/dynamic-dispatch.md:
```diff
@@ -162,3 +162,137 @@ main = Test.simplify
+     the [polyglot Interoperability](../polyglot/README.md) chapter.
+
+6. **There is no `Receiver`:**
+   - 6.1. We are looking either for a _module method_ or for a variable in the
```
- Not really a "method resolution"
- this is only influenced by "lexical scope" with all the defined symbols being available
- this should even yield an error during "compile time" - e.g. before executing any code

Akirathan replied:
In 5f6fdaa71cccb998f17f311e68d711c21d45d46a, it is clear that when there is no `Receiver`, it is no longer considered _method resolution_, but rather a _lexical scope symbol lookup_.

Commit:
Simplify docs for method resolution

JaroslavTulach committed:
Removing accidental changes

JaroslavTulach commented on docs/types/dynamic-dispatch.md:
```diff
@@ -162,3 +162,133 @@ main = Test.simplify
+
+- **eigen type** of a type `My_Type` is a type of type, usually written as
+  `My_Type.type`.
+  - Every type has an eigen type.
```
#### Observation

- there is always exactly one object being an instance of each eigen (own, vlastní) type
- module, type have each their own eigen type
- singleton type is its own eigen type

Akirathan replied:
Added in d841f70df14bd33d05ec46e9243954fd44c269f0

JaroslavTulach commented on docs/types/dynamic-dispatch.md:
```diff
@@ -162,3 +162,133 @@ main = Test.simplify
+- 1.4. If `Receiver` is a polyglot object, method resolution and invocation will
+  be handled according to the [polyglot interoperability](../polyglot/README.md)
+  rules.
+  - Polyglot object can be:
```
- This enumeration probably belongs to `../polyglot/README.md`
- e.g. no need to explain how to obtain polyglot object in this document
- reference to the other document would be more appropriate

Akirathan replied:
Improved in 8f7d18e3370d0ab9abb5ff3a76ed4b33bf764638. Added a reference to the polyglot section.

JaroslavTulach commented on docs/types/dynamic-dispatch.md:
```diff
@@ -162,3 +162,133 @@ main = Test.simplify
+2. **Look up symbol in the symbol table of the determined type:**
+
+- 2.1. If `symbol` is defined in the table, select it and stop.
+- 2.2. If parent type is present, repeat the process with the parent type.
```
- This is the first time you are using "parent type"
- there is no previous definition 
- maybe you want to expand the "terminology section"
- or create another "Reminiscence of OOP terminology" section

Akirathan replied:
Terminology for parent type and builtin type added in c6b9a3eaf5eec71d7e6f948704ada8000cde0915

JaroslavTulach commented on docs/types/dynamic-dispatch.md:
```diff
@@ -162,3 +162,133 @@ main = Test.simplify
+  - Every type has an implicit parent type `Any`.
+  - Except for `Number` which is a special builtin case with a deeper hierarchy.
+  - `Any` has no parent type.
+- 2.2. No parent type is present. Raise `No_Such_Method` panic and stop.
```
```suggestion
- 2.2. No parent type is present. Raise `No_Such_Method` panic and terminate.
```

JaroslavTulach commented on docs/types/dynamic-dispatch.md:
```diff
@@ -162,3 +162,133 @@ main = Test.simplify
+
+- 2.1. If `symbol` is defined in the table, select it and stop.
+- 2.2. If parent type is present, repeat the process with the parent type.
+  - Every type has an implicit parent type `Any`.
```
```suggestion
  - There is a special type called `Any` at the root of "class hierarchy"
  - Every type has an implicit parent type `Any`.
```

Akirathan replied:
Already explained in the terminology now.

JaroslavTulach commented on docs/types/dynamic-dispatch.md:
```diff
@@ -162,3 +162,133 @@ main = Test.simplify
+- 2.1. If `symbol` is defined in the table, select it and stop.
+- 2.2. If parent type is present, repeat the process with the parent type.
+  - Every type has an implicit parent type `Any`.
+  - Except for `Number` which is a special builtin case with a deeper hierarchy.
```
```suggestion
  - Except for builtin types `Float` and `Integer` 
     - their parent type is `Number` builtin type
     - e.g. numbers are forming a special case with a deeper hierarchy.
```

JaroslavTulach commented on docs/types/dynamic-dispatch.md:
```diff
@@ -162,3 +162,133 @@ main = Test.simplify
+
+### Lexical scope lookup
+
+- If `symbol` is defined in the current lexical scope, select it and stop.
```
- This is not really a "dynamic dispatch"
- but it is an interesting topic
- and complicated - e.g. term like _all transitively imported modules_ sounds scary
- I'd move it to its own _"lexical dispatch"_ document

JaroslavTulach commented on docs/types/dynamic-dispatch.md:
```diff
@@ -162,3 +162,133 @@ main = Test.simplify
+[Method Resolution](#method-resolution) process, that is defined as
+`method [self] [parName=[defaultValue]]* = <body expression>`.
+
+If `self` parameter is specified, we call it an _instance method_, otherwise, we
```
- I'd say this is wrong:
- `self` argument is always the first argument of each method
- at least for _each method_ that can be found in a _lookup table_ of a type
- this is an important step towards #11686

Akirathan replied:
> self argument is always the first argument of each method

It is, but from the engine perspective. From the engine perspective, every method has either a single `self` argument, or two `self` arguments. However, from the language perspective, you cannot declare `self` argument on, for example, module methods, neither can you specify `self` argument when you call such methods.

JaroslavTulach replied:
> two `self` arguments

- I don't understand why should there be two `self` arguments?
   - I heard about _type of self_ and `self` itself - I am not sure what that would be good for?
- I don't even know when has a method two such arguments?

> neither can you specify self argument when you call such module methods.

That would have little sense. `self` argument is type checked and there is only a singleton of a module associated type. E.g. specifying any other `self` to a module method would have to yield `Type_Error`.

Akirathan replied:
Closing as outdated.

JaroslavTulach commented on docs/types/dynamic-dispatch.md:
```diff
@@ -162,3 +162,133 @@ main = Test.simplify
+### Instance method call evaluation
+
+The method call expression in the format of `obj.method [[argName=]argValue]*`,
+where `obj` is an instance (value / atom) of type `T`, and `method` is an
```
```suggestion
where `obj` has type `T`, and `method` is an
```
this works for all types. Not just `Atom` instances.

Akirathan replied:
This whole section was reworded in 7590715c30731b5ea6002fb2768081e5cd5ea0f8. Resolving this comment as outdated.

JaroslavTulach commented on docs/types/dynamic-dispatch.md:
```diff
@@ -162,3 +162,133 @@ main = Test.simplify
+_instance method_ with `self` argument defined either on `T` or outside `T` as
+an _extension method_, the method evaluation process is as follows:
+
+- `self` argument cannot be specified explicitly
```
```suggestion
- `self` argument is set to `obj`by implicitly specifying`self=obj`
- possibly `self` could be overridden!? by explicitly specifying `self=xyz`
```

Akirathan replied:
Removed in 7590715c30731b5ea6002fb2768081e5cd5ea0f8

JaroslavTulach commented on docs/types/dynamic-dispatch.md:
```diff
@@ -162,3 +162,133 @@ main = Test.simplify
+an _extension method_, the method evaluation process is as follows:
+
+- `self` argument cannot be specified explicitly
+- Go to [Arguments evaluation](#argument-evaluation) to evaluate `argValue`s and
```
- argument evaluation is an interesting and rich topic
- we don't have problems with it
- we should just reference some other document or `CurryNode` code(?)
- rather than trying to define it here - that would blow the scope of this PR up tremendously

Akirathan replied:
argument evaluation section removed in 497f2ab526e944a9594bc9d6b719c2646c54c822

JaroslavTulach commented on docs/types/dynamic-dispatch.md:
```diff
@@ -162,3 +162,133 @@ main = Test.simplify
+- Go to [Arguments evaluation](#argument-evaluation) to evaluate `argValue`s and
+  bind them to `argName`s.
+
+### Static method call evaluation
```
```suggestion
### Static Method Invocation
```

JaroslavTulach commented on docs/types/dynamic-dispatch.md:
```diff
@@ -162,3 +162,133 @@ main = Test.simplify
+
+The method call expression in the format of
+`TypeOrModule.method [[argName=]argValue]*`, where `TypeOrModule` is either a
+type or a module, and `method` is a _static method_ defined on `TypeOrModule`.
```
- there are no static methods
- there are just "types (and eigen types) with their tables of methods"

Akirathan replied:
This whole section was completely reworded in 7590715c30731b5ea6002fb2768081e5cd5ea0f8. This comment is now outdated.

JaroslavTulach commented on docs/types/dynamic-dispatch.md:
```diff
@@ -162,3 +162,133 @@ main = Test.simplify
+Note that `TypeOrModule` does not have to be specified, if `method` is directly
+imported.
+
+The method evaluation process is as follows:
+
+1. **Determine whether `TypeOrModule` is a type or a module:**
+
+   - 1.1. If `TypeOrModule` is not `Any`, go to 2
+   - 1.2. If `TypeOrModule` is `Any`, go to 4
+
+2. **`TypeOrModule` is not `Any`:**
+
+   - 2.1. If `self` argument is specified explicitly, go to 3. `self` argument
+     value is either:
+     - Passed via `self` named argument.
+     - First positional argument.
+   - 2.2. If `self` argument is not specified explicitly, go to 4.
+
+3. **`self` is specified explicitly:**
+
+   - 3.1. `self` argument is specified explicitly.
+   - 3.2. Bind `self` argument to the provided value.
+   - 3.3. Go to [Arguments evaluation](#argument-evaluation) to evaluate rest of
+     the arguments.
+
+4. **`self` is not specified explicitly:**
+
+   - 4.1. `self` argument is not specified explicitly. This can happen only if
+     there are no arguments. Which results in a method reference.
+
+5. **`TypeOrModule` is `Any`:**
+   - 5.1. TODO ...
```
We should try to reformulate this...
```suggestion
```

- should (instance) method dispatch fail without finding a "symbol" of the method
- consider _Static Invocation Syntax_:

### Static Invocation Syntax

- if the receiver is type
- search its (_not its eigen type_!) chain of method table for a match
- if a method is found
- invoke it _without self_ argument bound to anything

Akirathan replied:
Reformulated according to the algorithm specified in https://github.com/enso-org/enso/pull/13962#pullrequestreview-3287447276 in 7590715c30731b5ea6002fb2768081e5cd5ea0f8

Commit:
Apply suggestion from @JaroslavTulach

Co-authored-by: Jaroslav Tulach <jaroslav.tulach@enso.org>

Commit:
Added some observations to eigen types

Commit:
Refer to polyglot readme.

Commit:
Add parent type and builtin type terminology

Commit:
Update docs/types/dynamic-dispatch.md

Co-authored-by: Jaroslav Tulach <jaroslav.tulach@enso.org>

Commit:
Reword and simplify.

Algorithm according to https://github.com/enso-org/enso/pull/13962#pullrequestreview-3287447276

Commit:
Remove argument evaluation section

JaroslavTulach commented on docs/types/dynamic-dispatch.md:
```diff
@@ -163,16 +163,32 @@ When invoking a method on _module object_ its _module static methods_ take
 primarily as a _container for module (static) methods_.
 
-## Method Resolution
+## Method invocation
```
```suggestion
## Method Invocation
```

JaroslavTulach commented on docs/types/dynamic-dispatch.md:
```diff
@@ -163,16 +163,32 @@ When invoking a method on _module object_ its _module static methods_ take
+    [eigen value of a matrix](https://en.wikipedia.org/wiki/Eigenvalues_and_eigenvectors).
 - **associated type** of a module `My_Module` is a type for the module
   - It is basically an eigen type for a module.
+- **builtin type** is a type annotated with `@Builtin_Type`.
```
- relevant question is: [What is a benefit of builtin type](https://github.com/enso-org/enso/issues/13664#issuecomment-3266102569)?
- unless I am mistaken we could easily run without `@Builtin_Type`
- I mean builtin type - not `@Builtin_Method` - that's still very much needed
- on and on, I'd rather leave _builtin types_ out of the method resolution document
- they don't have any special behavior to discuss - or do they?

Akirathan replied:
> relevant question is: https://github.com/enso-org/enso/issues/13664#issuecomment-3266102569?

That is indeed very relevant.

> they don't have any special behavior to discuss - or do they?

The only reason why I am mentioning builtin types in terminology section here, is because only builtin types may have different inheritance hierarchy. An example is later mentioned in [parent type](https://github.com/enso-org/enso/blob/41914f42ec66ae4b64c8b10215b79619e8bce9e3/docs/types/dynamic-dispatch.md#L189-L191).

JaroslavTulach commented on docs/types/dynamic-dispatch.md:
```diff
@@ -200,23 +216,37 @@ the `Receiver`, and then finds the method definition in its _symbol table_:
+  table.
+- 3.2. Such method is treated as if it's first parameter is named `self` and has
+  the default value of `Receiver`.
+- 3.3. The call site can specify named `self` argument
```
Super. Section three is short and exact.

JaroslavTulach commented on docs/types/dynamic-dispatch.md:
```diff
@@ -200,23 +216,37 @@ the `Receiver`, and then finds the method definition in its _symbol table_:
+- 4.2. Try to lookup the method on the `Receiver` itself
+- 4.3. If no method is found, raise `No_Such_Method` panic and stop.
+- 4.4. If a method is found, treat it as if it's first parameter is named `self`
+  and has no default value.
```
OK too.

JaroslavTulach commented on docs/types/dynamic-dispatch.md:
```diff
@@ -227,68 +257,64 @@ the `Receiver`, and then finds the method definition in its _symbol table_:
-- `self` argument cannot be specified explicitly
-- Go to [Arguments evaluation](#argument-evaluation) to evaluate `argValue`s and
-  bind them to `argName`s.
+### Example (a)
```
- Instead of examples in the documentation, put them into a test case and link from here. Maybe?
- Documentation full of examples has only one future: to get outdated.

Akirathan replied:
Valid point. I will definitely add such tests in https://github.com/enso-org/enso/pull/13962. This whole documentation is, in fact, the **desired behavior** that is being implemented in #13962.

Akirathan replied:
Added as a checklist point in #13962.

JaroslavTulach reviewed (approved):
Section 2, 3, 4 seem to be correct.

Commit:
Update docs/types/dynamic-dispatch.md

Co-authored-by: Jaroslav Tulach <jaroslav.tulach@enso.org>

Commit:
Merge branch 'develop' into wip/akirathan/docs-methods

Commit:
fmt

Commit:
typo

Commit:
Consistently ujse parameters and arguments

JaroslavTulach commented on docs/types/dynamic-dispatch.md:
```diff
@@ -162,3 +162,159 @@ main = Test.simplify
+- 2.2. If it is found, continue to 3.
+- 2.3. If it is not found, continue to 4.
+
+3. **Invoke the method with defaulted self**:
```
If there is no named argument `self=` specified, then we perform the invoke with self being set to receiver.

JaroslavTulach commented on docs/types/dynamic-dispatch.md:
```diff
@@ -162,3 +162,159 @@ main = Test.simplify
+  table.
+- 3.2. Such method is treated as if it's first parameter is named `self` and has
+  the default value of `Receiver`.
+- 3.3. The call site can specify named `self` argument
```
```suggestion
- 3.3. The call site cannot specify named `self` argument
```

JaroslavTulach commented on docs/types/dynamic-dispatch.md:
```diff
@@ -162,3 +162,159 @@ main = Test.simplify
+  the default value of `Receiver`.
+- 3.3. The call site can specify named `self` argument
+
+4. **Static invocation without bound self**:
```
If the first argument of the method is `self=`, then it is static method invocation syntax.

JaroslavTulach commented on docs/types/dynamic-dispatch.md:
```diff
@@ -162,3 +162,159 @@ main = Test.simplify
+
+4. **Static invocation without bound self**:
+
+- 4.1. Method was not found on Receiver's type, or its parent type.
```
```suggestion
- 4.1. If the first argument of the method is self=, then it is static method invocation syntax.
   - specifying `self=` argument as second or later is compile time error
```

JaroslavTulach commented on docs/types/dynamic-dispatch.md:
```diff
@@ -162,3 +162,159 @@ main = Test.simplify
+- 4.2. Try to lookup the method on the `Receiver` itself
+- 4.3. If no method is found, raise `No_Such_Method` panic and stop.
+- 4.4. If a method is found, treat it as if it's first parameter is named `self`
+  and has no default value.
```
```suggestion
  and has value specifying by the explicit `self=` argument
```

JaroslavTulach reviewed (approved):
Another update. Let's differentiate:
- instance invocation syntax has no `self=` named argument
- static syntax for invoking instance method has first argument `self=`
- the decision what to search is thus done on syntactic level 
- by presence/absence of the `self=` argument

Commit:
Update docs with the newest proposition

Akirathan commented:
https://github.com/enso-org/enso/pull/13962/commits/b15403ea0f09ec4407967ee926c4e6b554bd4bbb updates #13962 to match the new specification

JaroslavTulach committed:
Removing confusion by using different Text values

JaroslavTulach committed:
Checking the hierarchy of Any and its eigentype

JaroslavTulach commented on engine/runtime-integration-tests/src/test/java/org/enso/interpreter/node/expression/builtin/meta/TypeChainTest.java:
```diff
@@ -0,0 +1,104 @@
+    var all = raw.allTypes(ctx.ensoContext());
+
+    var exp1 = any.getEigentype();
+    assertArrayEquals("Any.type only", new Object[] {exp1}, all);
```
### ~~Wrong~~

- this is the snapshot of current type hierarchy
- is it good or bad, @Akirathan?
- this should mean that when typing `Any.to_text` to perform an _instance method invocation_
  -  the type of `Any` is `Any.type` - which is `Any`'s _eigentype_
  - if `to_text` is registered in `Any`'s _eigentype_ (**No, it is not!** it is registered in `Any` as instance method!)
  - then it will be found during _instance method invocation_
- thus `Any.to_text` should yield `"Any"`

### Correct

After a discussion with @Akirathan we concluded, that the current behavior is wrong and it should be
```suggestion
    var exp1 = any.getEigentype();
    var exp2 = ctx.ensoContext().getBuiltins().any();
    assertArrayEquals("Any.type and Any", new Object[] {exp1, exp2}, all);
```
that way the _instance method dispatch_ should work for `Any.to_text`

Akirathan replied:
It is bad. On develop, `to_text` is registered both on `Any` and on `Any.type`. But ideally, it should be registered only on `Any`.

Akirathan replied:
I believe that the correct semantics should be:
```
allTypes(Any.type) == [Any.type, Any]
allTypes(Any) == [Any]
allTypes(Text) == [Text, Any]
allTypes(Text.type) == [Text.type, Any]
allTypes(Integer) == [Integer, Number, Any]
```

Akirathan replied:
Tests updated in f9172f58aa7ea5d1efcf7342e14ca96878064d10 to reflect this semantics. The test for `allTypes(Any.type) == [Any.type, Any]` is failing, as expected.

JaroslavTulach commented on engine/runtime-integration-tests/src/test/java/org/enso/interpreter/node/expression/builtin/meta/TypeChainTest.java:
```diff
@@ -0,0 +1,104 @@
+
+    var exp1 = ctx.ensoContext().getBuiltins().text().getEigentype();
+    var exp2 = ctx.ensoContext().getBuiltins().any();
+    assertArrayEquals("Text.type and Any", new Object[] {exp1, exp2}, all);
```
- is this correct - in the [light of this test](https://github.com/enso-org/enso/pull/13961/files#r2401345452)?
- shouldn't the two elements in the array rather be `Type.type` and `Any.type`?
- why should a parent type of a value be `Any` and not `Any.type`?
- btw. if this is wrong, then it is _not a heritage_, but my mistake - I wrote the `allTypes` method

JaroslavTulach replied:
It is correct as [concluded here](https://github.com/enso-org/enso/pull/13961#discussion_r2401445009).

JaroslavTulach committed:
Type of Any.type is again Any.type

Commit:
Correct terminology - use preapplied argument instead of default

Akirathan commented:
I believe that as of https://github.com/enso-org/enso/pull/13961/commits/ed299f5be5bc3a4081bbd63a5268a375820ea920, this is really **the final specification** that should work for all our use cases. As demonstrated by the updated tests in https://github.com/enso-org/enso/pull/13962/commits/7b589378a49dbcb104f40166c064526a31ece732

JaroslavTulach commented:
> I believe that as of [ed299f5](https://github.com/enso-org/enso/commit/ed299f5be5bc3a4081bbd63a5268a375820ea920), this is really **the final specification** that should work for all our use cases. As demonstrated by the updated tests in [7b58937](https://github.com/enso-org/enso/commit/7b589378a49dbcb104f40166c064526a31ece732)

[This diff](https://github.com/enso-org/enso/pull/13962/files#r2403680299) is more important than https://github.com/enso-org/enso/pull/13961#issuecomment-3366485097 and it’s https://github.com/enso-org/enso/commit/7b589378a49dbcb104f40166c064526a31ece732 as it shows the diff against current behavior. Can we make it *less red*?

Akirathan commented:
> [This diff](https://github.com/enso-org/enso/pull/13962/files#r2403680299) is more important than https://github.com/enso-org/enso/pull/13961#issuecomment-3366485097 and it’s https://github.com/enso-org/enso/commit/7b589378a49dbcb104f40166c064526a31ece732 as it shows the diff against current behavior. Can we make it less red?

@JaroslavTulach Done. See https://github.com/enso-org/enso/pull/13962#discussion_r2405637786

Commit:
docs for singleton types

Commit:
Update test to reflect newest semantics

Commit:
Add tests for normal type and singleton type chains

Commit:
Update anyCHain test

Commit:
Implement correct semantics of Type.allTypes

JaroslavTulach commented on engine/runtime-integration-tests/src/test/java/org/enso/interpreter/node/expression/builtin/meta/TypeChainTest.java:
```diff
@@ -0,0 +1,169 @@
+    var exp1 = raw;
+    var exp2 = ctx.ensoContext().getBuiltins().any();
+    assertArrayEquals(
+        "allTypes(Singleton_Type.type) == [Singleton_Type.type, Any]",
```
Or is it
```suggestion
        "allTypes(Singleton_Type) == [Singleton_Type, Any]",
```
;-?

Akirathan replied:
I am really not sure about this one. See the line 159 with `assertThat("is eigen type", raw.isEigenType(), is(true))`. So I think it should be written as `Singleton_Type.type`.

Commit:
Merge branch 'develop' into wip/akirathan/docs-methods

Akirathan commented:
Changes introduced in [Type.fillInTypes](https://github.com/enso-org/enso/pull/13961/files#diff-fe7e602cb00fe7e5fe85ab1e94b7ee2c75cb79d1eb8fadc802f3eaca70474f00) necessary to make the newly introduced [TypeChainTests](https://github.com/enso-org/enso/pull/13961/files#diff-4dc127c7ebe31f3e61a1f95ca2eb7bbf4f81c9c862fd4ff444172e739a8bdc7f) green, breaks other stuff. For example in https://github.com/enso-org/enso/actions/runs/18308891816/job/52133745709?pr=13961#step:14:1114.

Can be reproduced with:
```
vec = [1,2,3]
vec.at 100 == Nothing
```
which, on this PR, returns `False`, but should return `Index_Out_Of_Bounds` error.

This is caused by the fact, that on develop `allTypes(Error) == [Error]`, but on this PR, `allTypes(Error) == [Error, Any]`. In turn, this means that an `UnresolvedSymbol("==").resolveFor(Error)` will resolve to `Any.==` which is incorrect.

`TypeChainTest` as well as changes in `Type` are necessary for #13962. Let's move it to that PR and keep this PR only about documentation.

-----------------

Moved in https://github.com/enso-org/enso/pull/13962/commits/1522b5096fc0c9760d4a0f3c7f11e8bd6e15e475 and https://github.com/enso-org/enso/pull/13962/commits/5a40ad8c354624ae5846b5ca1d9c0df55f4d9b81

Commit:
Revert Type to develop. Remove TypeChainTest.

These will be added to #13962.

JaroslavTulach commented:
> keep this PR only about documentation.

- (Unit) tests are essential part of documentation.
- I prefer to keep the test and just disable (or `@Ignore`) the parts that are known to diverge
- Leave comments with references to tasks that will fix the disabled parts

Commit:
Revert "Revert Type to develop. Remove TypeChainTest."

This reverts commit 14fa2971478d1ab78a914204cd08a515164634bb.

Commit:
NO changes to Type

Commit:
Ignore failing TypeChainTest

Akirathan commented:
@JaroslavTulach  Tests are again in this PR, with a single [ignore](feeb6ba86074d80ba5827f8c563d9fa99b0efb73).

Akirathan commented on engine/runtime-integration-tests/src/test/java/org/enso/interpreter/node/expression/builtin/meta/TypeChainTest.java:
```diff
@@ -0,0 +1,171 @@
+    assertArrayEquals(
+        "allTypes(Any.type) == [Any.type, Any]", new Object[] {anyEigenType, anyTypeExpected}, all);
+  }
+
```
TODO: Write test for `allTypes(Error) == [Error, Any]` (or `[Error]`??)

Akirathan replied:
Added in https://github.com/enso-org/enso/pull/13962/commits/e7efdf626139b4dc7b0fa9d26f64e04fa21b5ff9

JaroslavTulach commented on engine/runtime-integration-tests/src/test/java/org/enso/interpreter/node/expression/builtin/meta/TypeChainTest.java:
```diff
@@ -0,0 +1,171 @@
+  }
+
+  /** {@code allTypes(Any.type) == [Any.type, Any]} */
+  @Ignore("Will be addressed in #13939")
```
- being enabled by #14235
